### ISSUE TITLE
Prometheus can start off an agent WAL and vice versa

### ISSUE THREAD

**Issue by roidelapluie** [prometheus/prometheus#9615]
Currently, the agent can start off an agent WAL. I think we should somehow disallow that because the codepath to generate the WAL is not the same.

An alternative is to make sure that both WAL will remain compatible in the future.

**Comment by rfratto** [comment]
Could you explain this a little more? Is the concern that you might generate an agent WAL and then switch Prometheus to non-agent mode with the same directory?

**Comment by roidelapluie** [comment]
Yes but I actually think that the WAL has to be compatible for remot write, so we can probably ignore this?

However, what would happen if I run prometheus, then run a couple of hours as an agent, then run prometheus back? Could there be issues?

**Comment by rfratto** [comment]
Ah, interesting. The agent WAL should be compatible with TSDB's WAL with the exception of tombstones. The agent storage doesn't create them, and I don't know if that would cause any problems for TSDB. 

The deletion semantics are also going to be different: agent's WAL will more aggressively delete series in a way that the TSDB wouldn't expect. I agree this is another case of locking it down now and considering loosening the restriction later. 

As for how: maybe a touchfile that indicates agent/TSDB? That way, loading a WAL made from the other mode can return a failure. The user would be free to delete the file, but then whatever behavioral issues that might cause would be unsupported.

**Comment by roidelapluie** [comment]
I would have new magic codes for the agent Wal, and support both in the watcher

**Comment by codesome** [comment]
Both WALs would be compatible, and I think it is generally safer to allow switching with the same WAL. The deletion semantics does not change the format and replay characteristics of the WAL+Checkpoint. Trying to disallow switching feels like a wasted effort with no upside.

**Comment by codesome** [comment]
> However, what would happen if I run prometheus, then run a couple of hours as an agent, then run prometheus back? Could there be issues?

The major (and only?) difference is that I see, as Robert pointed out, will be how much data is left in WAL. Aka the deletion semantics.

**Comment by rfratto** [comment]
Should we maybe see if someone with a large existing TSDB is willing to toggle agent mode on/off a few times to see what happens? :)

**Comment by bboreham** [comment]
Reviewed at the bug-scrub meeting: seems nobody tried this or it works fine; either way we can close.
